numpyro/distributions/iaf.py

```diff
@@ -0,0 +1,84 @@
+# adapted from https://github.com/pyro-ppl/pyro/blob/dev/pyro/distributions/transforms/iaf.py
+
+from __future__ import absolute_import, division, print_function
+
+from numpyro.distributions.constraints import Transform, real
+from jax.lax import stop_gradient
+import jax.numpy as np
+
+
+def clamp_preserve_gradients(x, min, max):
+    return x + stop_gradient(np.clip(x, a_min=min, a_max=max) - x)
+
+
+class InverseAutoRegressiveTransform(Transform):
+    """
+    An implementation of Inverse Autoregressive Flow, using Eq (10) from Kingma et al., 2016,
+
+        :math:`\\mathbf{y} = \\mu_t + \\sigma_t\\odot\\mathbf{x}`
+
+    where :math:`\\mathbf{x}` are the inputs, :math:`\\mathbf{y}` are the outputs, :math:`\\mu_t,\\sigma_t`
+    are calculated from an autoregressive network on :math:`\\mathbf{x}`, and :math:`\\sigma_t>0`.
+
+    References
+
+    1. Improving Variational Inference with Inverse Autoregressive Flow [arXiv:1606.04934]
+    Diederik P. Kingma, Tim Salimans, Rafal Jozefowicz, Xi Chen, Ilya Sutskever, Max Welling
+    """
+    def __init__(self, autoregressive_nn, params, log_scale_min_clip=-5., log_scale_max_clip=3.):
+        """
+        :param autoregressive_nn: an autoregressive neural network whose forward call returns a real-valued
+            mean and logit-scale as a tuple
+        :param params: the parameters for the autoregressive neural network
+        :type list:
+        """
+        self.arn = autoregressive_nn
+        self.params = params
+        self.log_scale_min_clip = log_scale_min_clip
+        self.log_scale_max_clip = log_scale_max_clip
+
+    @property
+    def codomain(self):
```

Comment:
nit: Can be simplified to `codomain = real`, or we can also just leave it out and it will be correctly inherited from `Transform`.

Reply:
ok